# Phase 3 — Bayesian Structural Recommendation Engine

**Build order (start simple, add structure):** this notebook is step 1–2 — scope the DAG, then build a *baseline* hierarchical model with country partial-pooling on the treatment effect. Later steps add country/year effects, the mechanism layer, and counterfactual simulation.

## Step 1 — Structural DAG (scope)

The end-state wants a *mechanism breakdown* ("tax X%, switching Y%, ..."), which a reduced-form single coefficient can't give. The full causal chain is:

```
policy  ->  energy prices  ->  fuel mix  ->  emissions
 HAVE        MISSING            HAVE          HAVE
```

- **policy:** `has_tax`, `has_ets`, `tax_price_only`, `ets_price_only`, `fuel_subsidy_gdp`
- **energy prices:** NOT measured (we have *carbon* prices = the policy lever, not the retail energy price firms/households face). This link **collapses** to reduced-form.
- **fuel mix:** `fossil_pct_filled`, `renewable_pct`, `nuclear_pct`, `energy_per_capita`, and emissions split by fuel (`coal/gas/oil_co2_per_capita`).
- **emissions:** `co2_per_capita_future_trend` (3-yr forward).

Realistic skeleton: a two-link chain `policy -> fuel mix -> emissions`, decomposed via the **Kaya identity** `CO2/pop = (GDP/pop) x (Energy/GDP) x (CO2/Energy)` — fuel-switching is the last term (carbon intensity of energy at fixed demand), which separates the *behavioural* mechanism from the emissions accounting identity. Price node deferred (multi-week data project, extra identification assumption).

## Step 2 — Baseline hierarchical model

Country partial-pooling on the treatment effect. **Deliberately NOT yet causal** — no country intercepts / year effects yet (added next). Purpose: learn and verify the hierarchical machinery.

$$y_i \sim \mathrm{Normal}(\alpha + \beta_{c[i]}\cdot \mathrm{has\_tax}_i,\ \sigma)$$
$$\beta_c \sim \mathrm{Normal}(\mu, \tau)$$

Priors (weakly-informative; centered at no-effect so the data, not us, moves $\mu$):
`alpha ~ Normal(0, 0.5)`, `mu ~ Normal(0, 0.5)`, `tau ~ HalfNormal(0.5)`, `sigma ~ HalfNormal(1.0)`.
Locations -> Normal; scales (SDs) -> positive-only HalfNormal.

In [1]:
import os
os.environ['PYTENSOR_FLAGS'] = 'cxx='   # PyTensor C backend broken on Windows/py3.13; nutpie compiles via numba

import pandas as pd
import numpy as np
import pymc as pm
import arviz as az

In [2]:
df = pd.read_csv('../data/cleaned/final_analysis_data.csv')

country_idx, country_labels = pd.factorize(df['country'])   # row -> int 0..162 (realizes the c[i] lookup)
n_countries = len(country_labels)
y   = df['co2_per_capita_future_trend'].values
tax = df['has_tax'].values

print(f"obs: {len(df)}  countries: {n_countries}  treated rows (has_tax): {int(tax.sum())}")

obs: 4218  countries: 163  treated rows (has_tax): 261


**Non-centered hierarchy.** Writing `beta ~ Normal(mu, tau)` directly creates *Neal's funnel* — the region the betas can occupy depends on `tau`, so the sampler stalls (`tau` got ESS=36, R-hat=1.08 in the centered version). The fix: sample a standardized `z ~ Normal(0,1)` and rebuild `beta = mu + tau*z`. Statistically identical (`mu + tau*z ~ Normal(mu, tau)`), but the geometry is decoupled so the sampler glides.

In [3]:
with pm.Model() as baseline:
    alpha = pm.Normal('alpha', mu=0, sigma=0.5)
    mu    = pm.Normal('mu',    mu=0, sigma=0.5)
    tau   = pm.HalfNormal('tau',   sigma=0.5)
    sigma = pm.HalfNormal('sigma', sigma=1.0)

    z     = pm.Normal('z', mu=0, sigma=1, shape=n_countries)   # non-centered
    beta  = pm.Deterministic('beta', mu + tau * z)             # beta_c = mu + tau*z_c ~ Normal(mu, tau)

    mu_i  = alpha + beta[country_idx] * tax
    y_obs = pm.Normal('y_obs', mu=mu_i, sigma=sigma, observed=y)

In [4]:
with baseline:
    idata = pm.sample(draws=1000, tune=1000, chains=4, target_accept=0.9,
                      random_seed=42, nuts_sampler='nutpie', progressbar=False)

NUTS[nutpie]: [alpha, mu, tau, sigma, z]


**Diagnostics first, interpretation second.** Want R-hat ≈ 1.00 (chains agree) and ESS in the hundreds+ (enough effectively-independent draws), with 0 divergences.

In [5]:
print("Convergence (top-level params):")
print(az.summary(idata, var_names=['alpha', 'mu', 'tau', 'sigma'], round_to=4).to_string())
print(f"\nDivergences: {int(idata.sample_stats['diverging'].sum())}")

post_mu = idata.posterior['mu'].values.flatten()
print(f"mu posterior mean: {post_mu.mean():.4f}   P(mu < 0): {(post_mu < 0).mean():.3f}")

Convergence (top-level params):
         mean      sd  eti89_lb  eti89_ub    ess_bulk   ess_tail   r_hat  mcse_mean  mcse_sd
alpha -0.0079  0.0054   -0.0165    0.0007   9271.2929  3084.2262  1.0001     0.0001   0.0000
mu    -0.1640  0.0305   -0.2125   -0.1166   3277.9440  2948.1450  1.0009     0.0005   0.0004
tau    0.0896  0.0387    0.0252    0.1535    693.1524   631.8018  1.0054     0.0014   0.0011
sigma  0.3482  0.0037    0.3421    0.3541  10202.2983  2956.4865  1.0012     0.0000   0.0000

Divergences: 0
mu posterior mean: -0.1640   P(mu < 0): 1.000


**Read.** All R-hat ≈ 1.00, ESS in the hundreds-to-thousands, 0 divergences — converged. `mu` ≈ −0.16 with `P(mu<0)` ≈ 1.0 echoes the Phase-1/2 ATT.

**Do NOT report this as causal yet.** The model has no country intercepts and no year effects, so `mu` absorbs cross-country level differences and global time trends, not just the policy. Next step: add country intercepts `alpha_c` (same partial-pooling trick on the baseline level) + year effects to recover a DiD-style causal estimate, then anchor priors to Phase 1/2 and splice in the mechanism layer.

## Step 3 — Make it causal: two-way fixed effects

The baseline above is **not causal**: with a single global intercept, `mu` absorbs *pre-existing, time-invariant differences between taxed and untaxed countries* (taxed countries are richer, higher-governance, already declining) plus *global time trends*. That's the "Pooled OLS without FE → omitted variable bias" trap.

The fix is the same two-way fixed effects we used in the Phase-1 DiD (`C(country) + C(year)`), now expressed as **partial-pooled random effects**:

- **country intercepts** $\alpha_c$ remove time-invariant between-country confounders → $\beta$ is identified from *within-country* variation (same country, tax-on vs tax-off).
- **year effects** $\gamma_t$ remove time-varying confounders *common to all countries* (recessions, global renewables trend) → $\beta$ is net-of-the-common-time-path.

$$y_i \sim \mathrm{Normal}(\alpha_{c[i]} + \gamma_{t[i]} + \beta_{c[i]}\cdot \mathrm{has\_tax}_i,\ \sigma)$$
$$\alpha_c \sim \mathrm{Normal}(\mu_\alpha,\ \sigma_\alpha) \qquad \gamma_t \sim \mathrm{Normal}(0,\ \sigma_\gamma) \qquad \beta_c \sim \mathrm{Normal}(\mu,\ \tau)$$

**Identification anchor.** $\alpha_c$ gets a *free* grand mean $\mu_\alpha$ (it replaces the old global `alpha`), but $\gamma_t$'s mean is *pinned at 0*. Otherwise the additive degeneracy bites: add a constant $k$ to every $\alpha_c$ and subtract it from every $\gamma_t$ and the fit is identical — the sampler drifts along that ridge. Pinning the year-effect mean nails the level into the country intercepts.

**What this does NOT buy.** Two-way FE removes those two confounder types but not time-varying confounders *specific to taxing countries* — that residual is what **parallel trends** assumes away (stress-tested in Phase 1 via event study + Rambachan–Roth). FE gives the DiD estimand *conditional on* that assumption; it doesn't make it true.

In [6]:
year_idx, year_labels = pd.factorize(df['year'])   # row -> int 0..n_years-1
n_years = len(year_labels)
print(f"years: {n_years}  ({year_labels.min()}-{year_labels.max()})")

years: 26  (1996-2021)


In [7]:
with pm.Model() as did_model:
    # country baselines (partial-pooled FE) -- free grand mean mu_a replaces the old global alpha
    mu_a    = pm.Normal('mu_a', mu=0, sigma=0.5)
    sigma_a = pm.HalfNormal('sigma_a', sigma=0.5)
    z_a     = pm.Normal('z_a', mu=0, sigma=1, shape=n_countries)
    alpha   = pm.Deterministic('alpha', mu_a + sigma_a * z_a)        # alpha_c = mu_a + sigma_a*z_a

    # year effects (partial-pooled) -- mean pinned at 0 so the level lives in alpha
    sigma_g = pm.HalfNormal('sigma_g', sigma=0.5)
    z_g     = pm.Normal('z_g', mu=0, sigma=1, shape=n_years)
    gamma   = pm.Deterministic('gamma', sigma_g * z_g)               # mean 0 by construction

    # heterogeneous tax effect (unchanged from the baseline)
    mu    = pm.Normal('mu', mu=0, sigma=0.5)
    tau   = pm.HalfNormal('tau', sigma=0.5)
    z     = pm.Normal('z', mu=0, sigma=1, shape=n_countries)
    beta  = pm.Deterministic('beta', mu + tau * z)

    sigma = pm.HalfNormal('sigma', sigma=1.0)

    mu_i  = alpha[country_idx] + gamma[year_idx] + beta[country_idx] * tax
    y_obs = pm.Normal('y_obs', mu=mu_i, sigma=sigma, observed=y)

In [8]:
with did_model:
    idata_did = pm.sample(draws=1000, tune=1000, chains=4, target_accept=0.9,
                          random_seed=42, nuts_sampler='nutpie', progressbar=False)

NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu, tau, z, sigma]


In [9]:
print("Convergence (top-level params):")
print(az.summary(idata_did, var_names=['mu', 'tau', 'mu_a', 'sigma_a', 'sigma_g', 'sigma'],
                 round_to=4).to_string())
print(f"\nDivergences: {int(idata_did.sample_stats['diverging'].sum())}")

post_mu = idata_did.posterior['mu'].values.flatten()
print(f"mu posterior mean: {post_mu.mean():.4f}   P(mu < 0): {(post_mu < 0).mean():.3f}")

Convergence (top-level params):
           mean      sd  eti89_lb  eti89_ub   ess_bulk   ess_tail   r_hat  mcse_mean  mcse_sd
mu      -0.1214  0.0349   -0.1762   -0.0669  1877.6440  2354.8240  1.0013     0.0008   0.0006
tau      0.0864  0.0442    0.0173    0.1582   526.3971   611.6380  1.0066     0.0018   0.0014
mu_a    -0.0096  0.0152   -0.0336    0.0146   628.6630   974.9786  1.0037     0.0006   0.0004
sigma_a  0.1119  0.0083    0.0990    0.1258  1284.1217  1533.2119  1.0005     0.0002   0.0002
sigma_g  0.0551  0.0103    0.0405    0.0731  1049.5173  1550.9827  1.0085     0.0003   0.0003
sigma    0.3265  0.0036    0.3207    0.3324  5976.4660  2934.5931  1.0016     0.0000   0.0000

Divergences: 0
mu posterior mean: -0.1214   P(mu < 0): 0.999


**Read.** Converged (R-hat ≈ 1.00, ESS in the hundreds-to-thousands, 0 divergences). `mu` moved **−0.164 → −0.121**: the ~0.04 shift toward zero is the confounding the FE absorbed (`sigma_a` ≈ 0.11 — countries differ in baseline; `sigma_g` ≈ 0.055 — modest common year shifts), and `sigma` dropped because those structured effects explain real variance. What's left, **−0.121**, is the within-country, net-of-common-time DiD estimand.

**Validation:** −0.121 lands on the Phase-1/2 ATT (≈ −0.13). Frequentist two-way-FE DiD and this Bayesian partial-pooled model — different machines, same identification, same number.

**Caveats.** (1) `has_tax` only; Phase 2 found **ETS carries the effect** — next step adds `has_ets` as a second treatment, after which `mu` reads as "tax holding ETS fixed." (2) `mu` is the population mean of the heterogeneous `beta_c` (`tau` ≈ 0.086 spread); that spread is what will drive country-specific recommendations later.

## Step 4 — Second treatment: de-confound tax from ETS

`has_tax` and `has_ets` **co-occur** — most EU countries run both (EU ETS since 2005 + national carbon taxes). A tax-only model leaves ETS out, and because ETS also lowers emissions *and* is correlated with tax adoption, the tax coefficient **absorbs ETS's effect** (omitted-treatment bias). So Step 3's `mu = -0.121` is partly ETS leaking in.

This is the Phase-2 de-conflation of `post_carbon_tax` → clean `has_tax`/`has_ets`, now reproduced *inside* the Bayesian model by giving ETS its own coefficient:

$$y_i \sim \mathrm{Normal}\big(\alpha_{c[i]} + \gamma_{t[i]} + \beta_{c[i]}\cdot \mathrm{has\_tax}_i + \mu_{ets}\cdot \mathrm{has\_ets}_i,\ \sigma\big)$$

`mu_ets` is a **single pooled** coefficient for this increment (deferred: promote to heterogeneous `beta^{ets}_c`, and add the tax×ETS interaction — Phase 2's "both" cell). One new idea at a time: confirm the de-confounding echo first.

In [10]:
ets = df['has_ets'].values
print(f"treated rows -- has_tax: {int(tax.sum())}  has_ets: {int(ets.sum())}")

treated rows -- has_tax: 261  has_ets: 469


In [11]:
with pm.Model() as did_ets:
    # country baselines (partial-pooled FE) -- free grand mean
    mu_a    = pm.Normal('mu_a', mu=0, sigma=0.5)
    sigma_a = pm.HalfNormal('sigma_a', sigma=0.5)
    z_a     = pm.Normal('z_a', mu=0, sigma=1, shape=n_countries)
    alpha   = pm.Deterministic('alpha', mu_a + sigma_a * z_a)

    # year effects (partial-pooled) -- mean pinned at 0
    sigma_g = pm.HalfNormal('sigma_g', sigma=0.5)
    z_g     = pm.Normal('z_g', mu=0, sigma=1, shape=n_years)
    gamma   = pm.Deterministic('gamma', sigma_g * z_g)

    # heterogeneous tax effect
    mu    = pm.Normal('mu', mu=0, sigma=0.5)
    tau   = pm.HalfNormal('tau', sigma=0.5)
    z     = pm.Normal('z', mu=0, sigma=1, shape=n_countries)
    beta  = pm.Deterministic('beta', mu + tau * z)

    # pooled ETS effect (single coefficient -- promote to heterogeneous later)
    mu_ets = pm.Normal('mu_ets', mu=0, sigma=0.5)

    sigma = pm.HalfNormal('sigma', sigma=1.0)

    mu_i  = alpha[country_idx] + gamma[year_idx] + beta[country_idx] * tax + mu_ets * ets
    y_obs = pm.Normal('y_obs', mu=mu_i, sigma=sigma, observed=y)

In [12]:
with did_ets:
    idata_ets = pm.sample(draws=1000, tune=1000, chains=4, target_accept=0.9,
                          random_seed=42, nuts_sampler='nutpie', progressbar=False)

NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu, tau, z, mu_ets, sigma]


In [13]:
print("Convergence (top-level params):")
print(az.summary(idata_ets, var_names=['mu', 'mu_ets', 'tau', 'mu_a', 'sigma_a', 'sigma_g', 'sigma'],
                 round_to=4).to_string())
print(f"\nDivergences: {int(idata_ets.sample_stats['diverging'].sum())}")

post_mu  = idata_ets.posterior['mu'].values.flatten()
post_ets = idata_ets.posterior['mu_ets'].values.flatten()
print(f"mu (tax)  mean: {post_mu.mean():.4f}   P(<0): {(post_mu  < 0).mean():.3f}")
print(f"mu_ets    mean: {post_ets.mean():.4f}   P(<0): {(post_ets < 0).mean():.3f}")

Convergence (top-level params):
           mean      sd  eti89_lb  eti89_ub   ess_bulk   ess_tail   r_hat  mcse_mean  mcse_sd
mu      -0.0677  0.0357   -0.1230   -0.0102  1762.4679  2212.6627  1.0017     0.0009   0.0006
mu_ets  -0.1649  0.0236   -0.2030   -0.1272  1879.6809  2446.6048  1.0004     0.0005   0.0004
tau      0.0866  0.0432    0.0156    0.1557   445.6315   402.8621  1.0106     0.0019   0.0012
mu_a     0.0054  0.0139   -0.0166    0.0270   736.9452  1073.2280  1.0054     0.0005   0.0004
sigma_a  0.1062  0.0081    0.0938    0.1195  1547.1409  2112.6310  1.0012     0.0002   0.0001
sigma_g  0.0472  0.0092    0.0342    0.0628  1067.9706  1815.2944  1.0042     0.0003   0.0003
sigma    0.3254  0.0036    0.3196    0.3312  4929.2327  2885.5088  1.0012     0.0001   0.0000

Divergences: 0
mu (tax)  mean: -0.0677   P(<0): 0.971
mu_ets    mean: -0.1649   P(<0): 1.000


**Read.** Converged (R-hat ≈ 1.00, 0 divergences). Giving ETS its own coefficient nearly **halves the tax effect** and reproduces the Phase-2 verdict:

| | Step 3 (tax only) | Step 4 (tax + ETS) | Phase 2 (frequentist) |
|---|---|---|---|
| mu_tax | −0.121 | **−0.068** (P<0 = 0.97) | −0.056 (n.s.) |
| mu_ets | — | **−0.165** (P<0 = 1.00) | −0.13 (sig) |

The ~0.05 the tax coefficient shed is the ETS effect that had been leaking into it (omitted-treatment bias). `mu_ets` is strong and precise (more ETS rows, and pooled). **"Carbon pricing works; it's the ETS not the tax"** — the Phase-2 headline now falls out of the hierarchical model too. `mu_tax` reads honestly as the tax effect *holding ETS and the fixed effects fixed*: weak.

**Implication for the engine:** lean on the ETS lever for predicted reductions; treat a standalone carbon tax as a marginal mover.

**Deferred (backlog):** heterogeneous `beta^{ets}_c`; tax×ETS interaction (Phase 2's "both" cell, which was sub-additive).

## Step 5 — Anchor the priors (honestly)

So far every prior is weakly-informative at zero (`Normal(0, 0.5)`, `HalfNormal(0.5)`) — the model rediscovers everything from scratch. We *know* more than that (Phase 1/2: effects are modest and negative, ETS carries it). The prior is where that knowledge belongs.

**The trap.** Phase 1/2 estimated ATT ≈ −0.13 *from this same dataset*. Centering a prior on −0.13 and re-fitting on the same rows uses the data **twice** — once to build the prior, once in the likelihood — producing a posterior that is **artificially overconfident** (falsely narrow CIs). A prior must be *independent* of the likelihood data; our own same-data estimate is not.

**The honest fix.** Anchor to information that *isn't this dataset*:
- **Center stays at 0; tighten the scale.** The robust external knowledge is that effects are *modest* (an effect of 0.5 in these units is ~1.4σ — absurd). `Normal(0, 0.2)` encodes "small effects," which the whole carbon-pricing literature supports independently of our data. Center at 0 → the **sign is earned from data, not assumed**.
- **Regularize the hierarchy scales** (`τ`, `σ_α`, `σ_γ`) — tighter `τ` ⇒ stronger pooling ⇒ noisy per-country `β_c` shrink toward the mean. This is the regularization the recommendation engine needs.
- **Phase 1/2 becomes the *validation*, not the input** — if the posterior still lands near −0.13/−0.16 under an honest prior, that's a genuine consistency check.

We fit **loose vs anchored** side by side to see two things: (1) do the top-level estimates move? (they shouldn't — n=4,218 means data dominates the prior); (2) does the country-level `β_c` spread shrink? (it should — that's the regularization).

In [14]:
# countries ever taxed -- the only ones whose beta_c is informed by data
treated_countries = np.unique(country_idx[tax == 1])

def make_model(hp):
    with pm.Model() as m:
        mu_a    = pm.Normal('mu_a', mu=0, sigma=0.5)
        sigma_a = pm.HalfNormal('sigma_a', sigma=hp['s_a'])
        z_a     = pm.Normal('z_a', mu=0, sigma=1, shape=n_countries)
        alpha   = pm.Deterministic('alpha', mu_a + sigma_a * z_a)

        sigma_g = pm.HalfNormal('sigma_g', sigma=hp['s_g'])
        z_g     = pm.Normal('z_g', mu=0, sigma=1, shape=n_years)
        gamma   = pm.Deterministic('gamma', sigma_g * z_g)

        mu    = pm.Normal('mu', mu=0, sigma=hp['eff'])
        tau   = pm.HalfNormal('tau', sigma=hp['tau'])
        z     = pm.Normal('z', mu=0, sigma=1, shape=n_countries)
        beta  = pm.Deterministic('beta', mu + tau * z)

        mu_ets = pm.Normal('mu_ets', mu=0, sigma=hp['eff'])
        sigma  = pm.HalfNormal('sigma', sigma=hp['resid'])

        mu_i = alpha[country_idx] + gamma[year_idx] + beta[country_idx] * tax + mu_ets * ets
        pm.Normal('y_obs', mu=mu_i, sigma=sigma, observed=y)
    return m

def fit(m):
    with m:
        return pm.sample(draws=1000, tune=1000, chains=4, target_accept=0.9,
                         random_seed=42, nuts_sampler='nutpie', progressbar=False)

In [15]:
loose    = {'eff': 0.5, 'tau': 0.5, 's_a': 0.5, 's_g': 0.5, 'resid': 1.0}
anchored = {'eff': 0.2, 'tau': 0.2, 's_a': 0.2, 's_g': 0.1, 'resid': 0.5}

idata_loose    = fit(make_model(loose))
idata_anchored = fit(make_model(anchored))

NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu, tau, z, mu_ets, sigma]


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu, tau, z, mu_ets, sigma]


In [16]:
def report(name, idata):
    s = az.summary(idata, var_names=['mu', 'mu_ets', 'tau'], round_to=4)
    beta_means = idata.posterior['beta'].mean(('chain', 'draw')).values[treated_countries]
    div = int(idata.sample_stats['diverging'].sum())
    print(f"[{name}]  divergences={div}")
    print(s[['mean', 'sd', 'ess_bulk', 'r_hat']].to_string())
    print(f"  treated-country beta_c spread: sd={beta_means.std():.4f}  "
          f"range=[{beta_means.min():.3f}, {beta_means.max():.3f}]\n")

report('loose (step 4)', idata_loose)
report('anchored',       idata_anchored)

[loose (step 4)]  divergences=0
          mean      sd   ess_bulk   r_hat
mu     -0.0677  0.0357  1762.4679  1.0017
mu_ets -0.1649  0.0236  1879.6809  1.0004
tau     0.0866  0.0432   445.6315  1.0106
  treated-country beta_c spread: sd=0.0441  range=[-0.232, -0.014]

[anchored]  divergences=0
          mean      sd   ess_bulk   r_hat
mu     -0.0653  0.0339  1657.7089  1.0010
mu_ets -0.1632  0.0234  2180.5011  1.0007
tau     0.0822  0.0431   499.4770  1.0053
  treated-country beta_c spread: sd=0.0413  range=[-0.221, -0.018]



**Read.** Both converged, 0 divergences.

| | loose (step 4) | anchored | moved? |
|---|---|---|---|
| mu_tax | −0.068 (sd 0.036) | −0.065 (sd 0.034) | third decimal |
| mu_ets | −0.165 (sd 0.024) | −0.163 (sd 0.023) | third decimal |
| tau | 0.087 | 0.082 | barely |
| beta_c spread (sd / range) | 0.044 / [−0.232, −0.014] | 0.041 / [−0.221, −0.018] | **shrank** |

1. **Top-level estimates barely move** — tightening every effect prior 2.5× shifts `mu`/`mu_ets` in the third decimal. With n=4,218 the likelihood dominates the prior, so the headline (ETS −0.16, tax weak) is a property of the *data*, not the prior. Robustness check passed.
2. **Country-level spread shrinks** — `beta_c` tightened (sd 0.044 → 0.041, extremes pulled in). The tighter `tau` prior pools noisy per-country estimates harder. Modest here (the loose prior wasn't crazy; most treated countries have enough rows), but it's the regularization the engine needs and matters more for thin-data countries.
3. **No circularity, still validated** — we never fed −0.13 into the prior, yet `mu_ets` lands at −0.163, on the Phase-1/2 ATT. Agreement under an honest prior is a real consistency check, not a self-fulfilling one.

**The anchored prior set is our standing model going forward** (`idata_anchored`): same answer, honestly regularized, no double-counting.

## Step 6 — Structural mechanism layer (Kaya decomposition)

Everything so far is **reduced-form**: `policy → emissions`, one arrow. `mu_ets` tells us ETS lowers emissions but not *how*. The recommendation engine needs a **mechanism breakdown**, so we open the box: `policy → mechanism → emissions`.

**Scaffolding — the Kaya identity** decomposes emissions/capita into three multiplicative channels:

$$\frac{\text{CO}_2}{\text{pop}} = \underbrace{\frac{\text{GDP}}{\text{pop}}}_{\text{affluence (activity)}} \times \underbrace{\frac{\text{Energy}}{\text{GDP}}}_{\text{energy intensity (efficiency)}} \times \underbrace{\frac{\text{CO}_2}{\text{Energy}}}_{\text{carbon intensity (fuel switch)}}$$

The **fuel-switching** behavioural story lives entirely in **carbon intensity** — the channel a well-functioning carbon price *should* move (same energy, cleaner source = the "good reason" emissions fall). **Affluence** is the channel to *worry* about: if emissions fell because activity shrank or moved offshore (**leakage**), that's the "bad reason." Efficiency is a legitimate middle.

This is **Phase 1's confounder/mediator rule run in reverse**: the energy/GDP variables we *refused* to control for (because they're mediators that would block the mechanism) are exactly the pathways we now explicitly model.

**Why logs.** Kaya is multiplicative → additive in logs, so the 3-yr forward *log* change decomposes exactly: `dlog(CO2/pop) = dlog(A) + dlog(I) + dlog(K)`. (Our earlier outcome `co2_per_capita_future_trend` is a *level* trend, so it can't decompose — we switch to a log-change outcome here. Magnitudes differ from earlier steps; these are annualized log-points, ~percent per year.)

In [17]:
dfk = pd.read_csv('../data/cleaned/final_analysis_data.csv').sort_values(['country', 'year']).copy()

# Kaya level channels:  CO2/pop = A * I * K
dfk['kaya_A'] = dfk['gdp'] / dfk['population']                    # affluence  (GDP/pop)
dfk['kaya_I'] = dfk['energy_per_gdp']                            # energy intensity (Energy/GDP)
dfk['kaya_K'] = dfk['co2_per_capita'] / dfk['energy_per_capita']  # carbon intensity (CO2/Energy)
dfk['kaya_T'] = dfk['co2_per_capita']                           # total
for c in ['A', 'I', 'K', 'T']:
    dfk[f'log_{c}'] = np.log(dfk[f'kaya_{c}'])

# identity check (level): does A*I*K reproduce CO2/pop?
prod = dfk['kaya_A'] * dfk['kaya_I'] * dfk['kaya_K']
ok = dfk[['kaya_A', 'kaya_I', 'kaya_K', 'kaya_T']].notna().all(axis=1) & (dfk['kaya_T'] > 0)
rel = ((prod[ok] - dfk['kaya_T'][ok]) / dfk['kaya_T'][ok]).abs()
print(f"level identity A*I*K vs CO2/pop: max rel err={rel.max():.2e}  median={rel.median():.2e}")

level identity A*I*K vs CO2/pop: max rel err=5.92e-03  median=1.93e-04


In [18]:
# forward 3-yr annualized log change per channel (self-merge on country, year+3)
fwd = dfk[['country', 'year', 'log_A', 'log_I', 'log_K', 'log_T']].copy()
fwd['year'] = fwd['year'] - 3
fwd = fwd.rename(columns={f'log_{c}': f'log_{c}_p3' for c in ['A', 'I', 'K', 'T']})
dfk = dfk.merge(fwd, on=['country', 'year'], how='left')
for c in ['A', 'I', 'K', 'T']:
    dfk[f'd_{c}'] = (dfk[f'log_{c}_p3'] - dfk[f'log_{c}']) / 3.0

# decomposition closes? d_T == d_A + d_I + d_K
have = dfk[['d_A', 'd_I', 'd_K', 'd_T']].notna().all(axis=1)
resid = (dfk['d_T'][have] - (dfk['d_A'][have] + dfk['d_I'][have] + dfk['d_K'][have])).abs()
print(f"log decomposition d_T vs d_A+d_I+d_K: max resid={resid.max():.2e}  median={resid.median():.2e}")

log decomposition d_T vs d_A+d_I+d_K: max resid=3.53e-03  median=8.10e-05


In [19]:
# common mechanism sample: all channels + treatments present
mk = dfk[dfk[['d_A', 'd_I', 'd_K', 'd_T', 'has_tax', 'has_ets']].notna().all(axis=1)].copy()
c_idx, c_lab = pd.factorize(mk['country'])
t_idx, t_lab = pd.factorize(mk['year'])
nc, ny = len(c_lab), len(t_lab)
tax_k, ets_k = mk['has_tax'].values, mk['has_ets'].values
print(f"mechanism sample: {len(mk)} rows, {nc} countries")

def fit_channel(yvals):
    # anchored priors; country+year FE; BOTH treatments pooled for a clean decomposition
    with pm.Model():
        mu_a    = pm.Normal('mu_a', 0, 0.5)
        sigma_a = pm.HalfNormal('sigma_a', 0.2)
        alpha   = mu_a + sigma_a * pm.Normal('z_a', 0, 1, shape=nc)
        sigma_g = pm.HalfNormal('sigma_g', 0.1)
        gamma   = sigma_g * pm.Normal('z_g', 0, 1, shape=ny)
        mu_tax  = pm.Normal('mu_tax', 0, 0.2)
        mu_ets  = pm.Normal('mu_ets', 0, 0.2)
        sigma   = pm.HalfNormal('sigma', 0.2)
        mu_i = alpha[c_idx] + gamma[t_idx] + mu_tax * tax_k + mu_ets * ets_k
        pm.Normal('y_obs', mu=mu_i, sigma=sigma, observed=yvals)
        idata = pm.sample(draws=1000, tune=1000, chains=4, target_accept=0.9,
                          random_seed=42, nuts_sampler='nutpie', progressbar=False)
    return (idata.posterior['mu_tax'].values.flatten(),
            idata.posterior['mu_ets'].values.flatten(),
            int(idata.sample_stats['diverging'].sum()))

mechanism sample: 3728 rows, 163 countries


In [20]:
labels = {'T': 'TOTAL', 'K': 'carbon intensity (fuel switch)',
          'I': 'energy intensity (efficiency)', 'A': 'affluence (activity)'}
res = {}
for c in ['T', 'K', 'I', 'A']:
    t, e, div = fit_channel(mk[f'd_{c}'].values)
    res[c] = (t, e)
    print(f"[{labels[c]:32s}] div={div}  "
          f"mu_tax={t.mean():+.4f} (P<0={(t<0).mean():.2f})  "
          f"mu_ets={e.mean():+.4f} (P<0={(e<0).mean():.2f})")

st = sum(res[c][0].mean() for c in ['K', 'I', 'A'])
se = sum(res[c][1].mean() for c in ['K', 'I', 'A'])
print(f"\nclosure tax: total={res['T'][0].mean():+.4f}  K+I+A={st:+.4f}")
print(f"closure ets: total={res['T'][1].mean():+.4f}  K+I+A={se:+.4f}")

NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, mu_ets, sigma]


[TOTAL                           ] div=0  mu_tax=-0.0133 (P<0=0.99)  mu_ets=-0.0286 (P<0=1.00)


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, mu_ets, sigma]


[carbon intensity (fuel switch)  ] div=0  mu_tax=-0.0062 (P<0=0.86)  mu_ets=-0.0143 (P<0=1.00)


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, mu_ets, sigma]


[energy intensity (efficiency)   ] div=0  mu_tax=-0.0121 (P<0=0.97)  mu_ets=-0.0020 (P<0=0.67)


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, mu_ets, sigma]


[affluence (activity)            ] div=0  mu_tax=+0.0094 (P<0=0.00)  mu_ets=-0.0093 (P<0=1.00)

closure tax: total=-0.0133  K+I+A=-0.0090
closure ets: total=-0.0286  K+I+A=-0.0256


**Read (annualized log-points; ×3 ≈ window %, log-point ≈ percent).**

| channel | mu_ets/yr | mu_tax/yr |
|---|---|---|
| TOTAL | −0.0286 (P<0=1.00) | −0.0133 (P<0=0.99) |
| ✅ carbon intensity (fuel switch) | **−0.0143** (P<0=1.00) | −0.0062 (P<0=0.86) |
| 🟡 energy intensity (efficiency) | −0.0020 (P<0=0.67) | −0.0121 (P<0=0.97) |
| ⚠️ affluence (activity) | −0.0093 (P<0=1.00) | +0.0094 (P<0=0.00) |

- **ETS:** ~half its effect is real **fuel-switching** (carbon intensity, strong) — the good channel — but ~a third rides on **lower activity** (affluence) — the worry channel (leakage / suppressed activity). Efficiency ≈ 0. Over a 3-yr window: ≈ −8% total ≈ −4% decarbonization + −3% activity. The engine should flag the activity share.
- **Carbon tax:** weak total, and what it does runs through **efficiency**, not fuel-switching (carbon intensity uncertain, P<0=0.86); its affluence channel is *positive* (taxing countries grew, masking the effect). Mechanistic echo of Phase 2's null outcome-sensitivity.

**Caveats.** (1) Closure ~85–90%, not exact (OLS would close to machine precision; priors + partial-pooled FE shrink each channel independently). Trust the *shape*, not the last decimal. (2) The affluence channel is an *association* within the DiD, not a proven causal mechanism — mediation assumes no mediator–outcome confounding (strong). (3) Pooled treatments (heterogeneity set aside for a clean split).

## Step 7 — Counterfactual recommendation engine (with mechanism breakdown)

This is the payoff: *"Country X, adopt policy P → predicted change + CI + mechanism breakdown."*

**The mechanic.** A recommendation is a **counterfactual contrast** — the same country with the policy on vs off, differenced. Our posterior is ~4,000 self-consistent draws; we difference **within each draw**, so the country baseline `alpha_c` and year effect `gamma_t` (identical across the two worlds) cancel. The contrast for adopting the tax collapses to the posterior of `beta_c`; for ETS, `mu_ets`; for both, `beta_c + mu_ets`. **The recommendation is literally a posterior summary** — uncertainty propagated end-to-end, no delta method or bootstrap.

**Honest CI by construction (partial pooling).** For a country we observed taxed, `beta_c` is data-informed (tighter); for a never-taxed country the model never learned its `beta_c`, so it reverts to the population `Normal(mu, tau)` → automatically **wider CI**. "We're extrapolating" shows up as a wider interval on its own.

**Mechanism breakdown.** We fit the *same* anchored, heterogeneous-tax engine on each Kaya channel (`d_T`, `d_K`, `d_I`, `d_A`), so each recommendation also splits into fuel-switching / efficiency / activity. (Reuses the Step-6 mechanism sample `mk` and indices.)

In [21]:
def fit_engine(yvals):
    # anchored priors; country+year FE; heterogeneous tax (beta_c) + pooled ETS
    with pm.Model():
        mu_a    = pm.Normal('mu_a', 0, 0.5)
        sigma_a = pm.HalfNormal('sigma_a', 0.2)
        alpha   = mu_a + sigma_a * pm.Normal('z_a', 0, 1, shape=nc)
        sigma_g = pm.HalfNormal('sigma_g', 0.1)
        gamma   = sigma_g * pm.Normal('z_g', 0, 1, shape=ny)
        mu      = pm.Normal('mu', 0, 0.2)
        tau     = pm.HalfNormal('tau', 0.2)
        beta    = pm.Deterministic('beta', mu + tau * pm.Normal('z', 0, 1, shape=nc))
        mu_ets  = pm.Normal('mu_ets', 0, 0.2)
        sigma   = pm.HalfNormal('sigma', 0.2)
        mu_i = alpha[c_idx] + gamma[t_idx] + beta[c_idx] * tax_k + mu_ets * ets_k
        pm.Normal('y_obs', mu=mu_i, sigma=sigma, observed=yvals)
        idata = pm.sample(draws=1000, tune=1000, chains=4, target_accept=0.9,
                          random_seed=42, nuts_sampler='nutpie', progressbar=False)
    return (idata.posterior['beta'].values.reshape(-1, nc),
            idata.posterior['mu_ets'].values.flatten())

CH = {'T': 'TOTAL', 'K': 'fuel-switching (carbon int.)',
      'I': 'efficiency (energy int.)', 'A': 'activity (affluence)'}
beta_eng, ets_eng = {}, {}
for c in CH:
    beta_eng[c], ets_eng[c] = fit_engine(mk[f'd_{c}'].values)
    print(f"fit channel {c}")

NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu, tau, z, mu_ets, sigma]


fit channel T


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu, tau, z, mu_ets, sigma]


fit channel K


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu, tau, z, mu_ets, sigma]


fit channel I


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu, tau, z, mu_ets, sigma]


fit channel A


In [22]:
pos_of   = {c: i for i, c in enumerate(c_lab)}
tax_years = mk.groupby('country')['has_tax'].sum()

def contrast(country, use_tax, use_ets, chan):
    p = pos_of[country]
    return beta_eng[chan][:, p] * use_tax + ets_eng[chan] * use_ets

def recommend(country, use_tax=True, use_ets=True):
    policy = ' + '.join([p for p, on in [('tax', use_tax), ('ETS', use_ets)] if on]) or 'nothing'
    tot = contrast(country, use_tax, use_ets, 'T')
    informed = 'data-informed' if tax_years.get(country, 0) > 0 else 'EXTRAPOLATED (never taxed)'
    print(f"\n=== {country} | adopt {policy} | tax effect: {informed} ===")
    print(f"  TOTAL  {tot.mean():+.4f}/yr  90% CI [{np.percentile(tot,5):+.4f}, {np.percentile(tot,95):+.4f}]  "
          f"P(reduce)={(tot<0).mean():.2f}  (~{tot.mean()*3*100:+.1f}% over 3yr)")
    for c in ['K', 'I', 'A']:
        d = contrast(country, use_tax, use_ets, c)
        sh = 100 * d.mean() / tot.mean() if abs(tot.mean()) > 1e-9 else float('nan')
        print(f"    {CH[c]:30s} {d.mean():+.4f}  [{np.percentile(d,5):+.4f}, {np.percentile(d,95):+.4f}]  share={sh:4.0f}%")
    ksum = sum(contrast(country, use_tax, use_ets, c).mean() for c in ['K', 'I', 'A'])
    print(f"    channels close ~{100*ksum/tot.mean():.0f}% of total")

In [23]:
known = tax_years.sort_values(ascending=False).index[0]   # most-taxed country
never = tax_years[tax_years == 0].index[0]                # a never-adopter

recommend(known, use_tax=True,  use_ets=True)
recommend(known, use_tax=False, use_ets=True)
recommend(known, use_tax=True,  use_ets=False)
recommend(never, use_tax=True,  use_ets=True)   # note the wider tax CI (extrapolated)


=== Sweden | adopt tax + ETS | tax effect: data-informed ===
  TOTAL  -0.0431/yr  90% CI [-0.0598, -0.0274]  P(reduce)=1.00  (~-12.9% over 3yr)
    fuel-switching (carbon int.)   -0.0203  [-0.0339, -0.0066]  share=  47%
    efficiency (energy int.)       -0.0147  [-0.0306, +0.0009]  share=  34%
    activity (affluence)           -0.0008  [-0.0117, +0.0091]  share=   2%
    channels close ~83% of total

=== Sweden | adopt ETS | tax effect: data-informed ===
  TOTAL  -0.0288/yr  90% CI [-0.0357, -0.0219]  P(reduce)=1.00  (~-8.6% over 3yr)
    fuel-switching (carbon int.)   -0.0142  [-0.0211, -0.0072]  share=  49%
    efficiency (energy int.)       -0.0019  [-0.0096, +0.0060]  share=   7%
    activity (affluence)           -0.0094  [-0.0136, -0.0050]  share=  32%
    channels close ~88% of total

=== Sweden | adopt tax | tax effect: data-informed ===
  TOTAL  -0.0142/yr  90% CI [-0.0306, +0.0008]  P(reduce)=0.94  (~-4.3% over 3yr)
    fuel-switching (carbon int.)   -0.0061  [-0.0192, +0.

**Read.** A real engine output for the most-taxed country: *"adopt both → ≈ −13% over 3 yr, 90% CI tight, near-certain; ≈half fuel-switching, the rest efficiency."* The policy comparison falls straight out:

- **ETS** is the confident lever (P(reduce)=1.00, CI entirely below 0) for every country; its split is ≈half fuel-switching + ≈third activity (matches Step-6).
- **Tax alone** is weak (P≈0.94, CI crosses 0) and runs through *efficiency*, not fuel-switching — the project's verdict, now at recommendation level.
- **Adopting both** nets the activity channel down because the tax's *positive* activity component partly offsets ETS's negative one — surfaced automatically.
- The **never-adopter** gets a visibly **wider tax CI** (reverts to the population distribution) and is flagged EXTRAPOLATED — honest by construction.

**Caveats carried into every recommendation:** (1) channels close ~73–88%, not exact — trust the shape; (2) shares destabilize when the total is small/uncertain (e.g. tax-only) — read shares only for confident recommendations; (3) ETS is still **pooled** (country-invariant) — the backlog `heterogeneous beta_ets` item is what lets the engine differentiate ETS effectiveness across countries; (4) the activity channel is an *association*, not a proven causal mechanism.

**Remaining for the engine:** heterogeneous `beta_ets`, tax×ETS interaction (backlog), then the **Streamlit dashboard** (build-fast) wrapping `recommend()` in country-picker + policy sliders.

## Step 8 — Backlog refinements folded into the engine

Two upgrades to the engine, both checked for identification first:

- **Heterogeneous `beta_ets`** (was pooled): 29 ever-ETS countries, median 14 ETS-years each → **well-identified**. Lets the engine differentiate ETS across countries (though `tau_ets` turns out small → only mild differentiation).
- **tax×ETS interaction `delta`** (the "both" cell): policy-cell counts are none=3264, tax-only=83, ETS-only=277, both=105 → **identifiable but weak** (the small tax-only + both cells are all that separate `delta`). Expect a wide-ish posterior, consistent with Phase 2 calling this "suggestive, data-limited."

$$y_i = \alpha_{c} + \gamma_{t} + \beta^{tax}_{c}\,\mathrm{tax}_i + \beta^{ets}_{c}\,\mathrm{ets}_i + \delta\,(\mathrm{tax}_i\!\cdot\!\mathrm{ets}_i) + \varepsilon_i$$

**This supersedes the Step-7 engine** (pooled ETS, additive). Same anchored priors; `delta` pooled.

In [24]:
both_k = tax_k * ets_k

def fit_engine2(yvals):
    with pm.Model():
        mu_a    = pm.Normal('mu_a', 0, 0.5)
        sigma_a = pm.HalfNormal('sigma_a', 0.2)
        alpha   = mu_a + sigma_a * pm.Normal('z_a', 0, 1, shape=nc)
        sigma_g = pm.HalfNormal('sigma_g', 0.1)
        gamma   = sigma_g * pm.Normal('z_g', 0, 1, shape=ny)
        mu_tax  = pm.Normal('mu_tax', 0, 0.2)
        tau_tax = pm.HalfNormal('tau_tax', 0.2)
        beta_tax = pm.Deterministic('beta_tax', mu_tax + tau_tax * pm.Normal('z_tax', 0, 1, shape=nc))
        mu_ets  = pm.Normal('mu_ets', 0, 0.2)
        tau_ets = pm.HalfNormal('tau_ets', 0.2)             # NEW: ETS now heterogeneous
        beta_ets = pm.Deterministic('beta_ets', mu_ets + tau_ets * pm.Normal('z_ets', 0, 1, shape=nc))
        delta   = pm.Normal('delta', 0, 0.2)               # NEW: tax x ETS interaction
        sigma   = pm.HalfNormal('sigma', 0.2)
        mu_i = alpha[c_idx] + gamma[t_idx] + beta_tax[c_idx]*tax_k + beta_ets[c_idx]*ets_k + delta*both_k
        pm.Normal('y_obs', mu=mu_i, sigma=sigma, observed=yvals)
        idata = pm.sample(draws=1000, tune=1000, chains=4, target_accept=0.95,
                          random_seed=42, nuts_sampler='nutpie', progressbar=False)
    return (idata.posterior['beta_tax'].values.reshape(-1, nc),
            idata.posterior['beta_ets'].values.reshape(-1, nc),
            idata.posterior['delta'].values.flatten(),
            int(idata.sample_stats['diverging'].sum()))

CH = {'T': 'TOTAL', 'K': 'fuel-switching (carbon int.)',
      'I': 'efficiency (energy int.)', 'A': 'activity (affluence)'}
bt, be, dl = {}, {}, {}
for c in CH:
    bt[c], be[c], dl[c], div = fit_engine2(mk[f'd_{c}'].values)
    print(f"fit channel {c}  div={div}")

NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma]


C:\Users\willn\OneDrive\Documents\github\CarbonTaxModellindAndRecommendation\.venv\Lib\site-packages\pytensor\tensor\rewriting\elemwise.py:1034: UserWarning: Loop fusion failed because the resulting node would exceed the kernel argument limit.
  warn(


fit channel T  div=0


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma]


C:\Users\willn\OneDrive\Documents\github\CarbonTaxModellindAndRecommendation\.venv\Lib\site-packages\pytensor\tensor\rewriting\elemwise.py:1034: UserWarning: Loop fusion failed because the resulting node would exceed the kernel argument limit.
  warn(


fit channel K  div=0


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma]


C:\Users\willn\OneDrive\Documents\github\CarbonTaxModellindAndRecommendation\.venv\Lib\site-packages\pytensor\tensor\rewriting\elemwise.py:1034: UserWarning: Loop fusion failed because the resulting node would exceed the kernel argument limit.
  warn(


fit channel I  div=0


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma]


C:\Users\willn\OneDrive\Documents\github\CarbonTaxModellindAndRecommendation\.venv\Lib\site-packages\pytensor\tensor\rewriting\elemwise.py:1034: UserWarning: Loop fusion failed because the resulting node would exceed the kernel argument limit.
  warn(


fit channel A  div=0


In [25]:
def contrast(country, use_tax, use_ets, chan):
    p = pos_of[country]
    return bt[chan][:, p]*use_tax + be[chan][:, p]*use_ets + dl[chan]*(use_tax*use_ets)

def recommend(country, use_tax=True, use_ets=True):
    policy = ' + '.join([p for p, on in [('tax', use_tax), ('ETS', use_ets)] if on]) or 'nothing'
    tot = contrast(country, use_tax, use_ets, 'T')
    info = 'data-informed' if tax_years.get(country, 0) > 0 else 'EXTRAPOLATED (never taxed)'
    print(f"\n=== {country} | adopt {policy} | tax effect: {info} ===")
    print(f"  TOTAL {tot.mean():+.4f}/yr  90% CI [{np.percentile(tot,5):+.4f}, {np.percentile(tot,95):+.4f}]  "
          f"P(reduce)={(tot<0).mean():.2f}  (~{tot.mean()*3*100:+.1f}% over 3yr)")
    for c in ['K', 'I', 'A']:
        d = contrast(country, use_tax, use_ets, c)
        sh = 100 * d.mean() / tot.mean() if abs(tot.mean()) > 1e-9 else float('nan')
        print(f"    {CH[c]:30s} {d.mean():+.4f}  share={sh:4.0f}%")

known = tax_years.sort_values(ascending=False).index[0]
never = tax_years[tax_years == 0].index[0]
for ut, ue in [(True, True), (False, True), (True, False)]:
    recommend(known, ut, ue)
recommend(never, True, True)

d = dl['T']
print(f"\ninteraction delta (total) = {d.mean():+.4f}  90% CI [{np.percentile(d,5):+.4f}, {np.percentile(d,95):+.4f}]  P(>0)={(d>0).mean():.2f}")
mtax = bt['T'][:, pos_of[known]]
print(f"tax alone = {mtax.mean():+.4f} | tax on top of ETS = {(mtax + d).mean():+.4f}  (redundancy)")


=== Sweden | adopt tax + ETS | tax effect: data-informed ===
  TOTAL -0.0377/yr  90% CI [-0.0581, -0.0186]  P(reduce)=1.00  (~-11.3% over 3yr)
    fuel-switching (carbon int.)   -0.0213  share=  56%
    efficiency (energy int.)       -0.0108  share=  29%
    activity (affluence)           +0.0010  share=  -3%

=== Sweden | adopt ETS | tax effect: data-informed ===
  TOTAL -0.0321/yr  90% CI [-0.0444, -0.0200]  P(reduce)=1.00  (~-9.6% over 3yr)
    fuel-switching (carbon int.)   -0.0139  share=  43%
    efficiency (energy int.)       -0.0035  share=  11%
    activity (affluence)           -0.0142  share=  44%

=== Sweden | adopt tax | tax effect: data-informed ===
  TOTAL -0.0227/yr  90% CI [-0.0413, -0.0049]  P(reduce)=0.98  (~-6.8% over 3yr)
    fuel-switching (carbon int.)   -0.0050  share=  22%
    efficiency (energy int.)       -0.0175  share=  77%
    activity (affluence)           +0.0048  share= -21%

=== Afghanistan | adopt tax + ETS | tax effect: EXTRAPOLATED (never taxed) ==

**Read — the tax is redundant, not weak.**

`delta = +0.017` (P(>0)=0.96): running both instruments is **sub-additive**. Decomposing the cells (most-taxed country):
- tax **alone** ≈ −0.023 (P(reduce)=0.98) — on its own the tax *does* cut emissions, confidently.
- tax **on top of ETS** ≈ −0.006 — once carbon is already priced via ETS, adding a tax buys almost nothing.

So the earlier "tax is weak" was the both-cell dragging `mu_tax` toward zero; with `delta` separating it, the standalone tax is real but **largely redundant with ETS** — carbon-pricing instruments behave like **substitutes, not complements**. Policy upshot: don't double-instrument expecting double the cut. The engine now reflects this — "both" dropped (−0.043 → −0.038) because it no longer adds the two effects naively.

**Caveats (carried into the recommendations):**
1. **Standalone `mu_tax` is fragile** — identified mostly from the 83-row tax-only cell (non-EU / pre-ETS Nordic years), which may decarbonize for non-tax reasons. Treat "tax alone −0.023" as the *least* robust number here; the redundancy (`delta`) and ETS strength are firmer.
2. **`tau_ets` is small** → ETS effectiveness is fairly homogeneous across the observed (mostly EU) countries; heterogeneous `beta_ets` gives only mild country-differentiation, and ETS recommendations for non-EU countries are extrapolation.
3. `delta` is pooled (one interaction for all countries/channels); channel closure still ~80%; the activity channel remains an association, not a proven mechanism.

**Engine complete.** `recommend(country, use_tax, use_ets)` → total Δ + 90% CI + P(reduce) + mechanism breakdown, with honest extrapolation flags. Remaining Phase 3: the **Streamlit dashboard** (build-fast) wrapping this.

## Step 9 — Validation (PPC, LOO) and robustness (Student-t)

Convergence (R-hat/ESS/divergences) only says the sampler worked — **not that the model fits**. Two tools close that gap, then we fix what they find.

- **Posterior predictive checks (PPC):** simulate datasets from the posterior and compare to the real data. A *Bayesian p-value* near 0.5 = the model reproduces that feature; near 0/1 = systematic misfit.
- **LOO-CV (PSIS):** out-of-sample predictive accuracy from a single fit. `elpd` (higher=better) compares models; `Pareto-k > 0.7` flags where the approximation is unreliable.

In [26]:
# total-model variants for validation: Normal+FE, Normal+no-FE, Student-t+FE
def build_total(likelihood, with_fe):
    with pm.Model() as m:
        if with_fe:
            alpha = pm.Normal('mu_a', 0, 0.5) + pm.HalfNormal('sigma_a', 0.2) * pm.Normal('z_a', 0, 1, shape=nc)
            gamma = pm.HalfNormal('sigma_g', 0.1) * pm.Normal('z_g', 0, 1, shape=ny)
            base = alpha[c_idx] + gamma[t_idx]
        else:
            base = pm.Normal('a', 0, 0.5)
        beta_tax = pm.Normal('mu_tax', 0, 0.2) + pm.HalfNormal('tau_tax', 0.2) * pm.Normal('z_tax', 0, 1, shape=nc)
        beta_ets = pm.Normal('mu_ets', 0, 0.2) + pm.HalfNormal('tau_ets', 0.2) * pm.Normal('z_ets', 0, 1, shape=nc)
        delta = pm.Normal('delta', 0, 0.2)
        sigma = pm.HalfNormal('sigma', 0.2)
        mu_i = base + beta_tax[c_idx]*tax_k + beta_ets[c_idx]*ets_k + delta*both_k
        if likelihood == 'normal':
            pm.Normal('y_obs', mu=mu_i, sigma=sigma, observed=mk['d_T'].values)
        else:
            pm.StudentT('y_obs', nu=pm.Gamma('nu', 2, 0.1), mu=mu_i, sigma=sigma, observed=mk['d_T'].values)
    return m

def fit_ll(m):
    with m:
        idata = pm.sample(draws=1000, tune=1000, chains=4, target_accept=0.95,
                          random_seed=42, nuts_sampler='nutpie', progressbar=False)
        pm.compute_log_likelihood(idata, progressbar=False)
    return idata

id_normal = fit_ll(build_total('normal', True))
id_nofe   = fit_ll(build_total('normal', False))
id_studt  = fit_ll(build_total('studentt', True))

NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma]


C:\Users\willn\OneDrive\Documents\github\CarbonTaxModellindAndRecommendation\.venv\Lib\site-packages\pytensor\tensor\rewriting\elemwise.py:1034: UserWarning: Loop fusion failed because the resulting node would exceed the kernel argument limit.
  warn(


NUTS[nutpie]: [a, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma]


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma, nu]


C:\Users\willn\OneDrive\Documents\github\CarbonTaxModellindAndRecommendation\.venv\Lib\site-packages\pytensor\tensor\rewriting\elemwise.py:1034: UserWarning: Loop fusion failed because the resulting node would exceed the kernel argument limit.
  warn(


In [27]:
def ppc_table(model, idata, yobs):
    with model:
        pp = pm.sample_posterior_predictive(idata, progressbar=False, random_seed=42)
    yrep = pp.posterior_predictive['y_obs'].values.reshape(-1, len(yobs))
    out = {}
    for name, fn in [('mean', np.mean), ('sd', np.std), ('q05', lambda a: np.percentile(a, 5)),
                     ('q95', lambda a: np.percentile(a, 95)), ('min', np.min), ('max', np.max)]:
        rep = np.array([fn(yrep[i]) for i in range(0, yrep.shape[0], 8)])
        out[name] = (fn(yobs), rep.mean(), (rep >= fn(yobs)).mean())
    return out

yT = mk['d_T'].values
print("PPC Bayesian p-values (near 0.5 = good;  near 0/1 = misfit)")
print(f"{'stat':5s} {'obs':>9s} {'Normal-p':>9s} {'StudentT-p':>11s}")
ppn, ppt = ppc_table(build_total('normal', True), id_normal, yT), ppc_table(build_total('studentt', True), id_studt, yT)
for s in ['mean', 'sd', 'q05', 'q95', 'min', 'max']:
    print(f"{s:5s} {ppn[s][0]:+9.4f} {ppn[s][2]:9.2f} {ppt[s][2]:11.2f}")

Sampling: [y_obs]


PPC Bayesian p-values (near 0.5 = good;  near 0/1 = misfit)
stat        obs  Normal-p  StudentT-p


Sampling: [y_obs]


mean    +0.0096      0.52        0.34
sd      +0.0597      0.54        0.99
q05     -0.0743      0.00        0.19
q95     +0.1067      0.69        0.00
min     -0.3841      1.00        0.00
max     +0.4843      0.00        0.99


In [28]:
# LOO: FE earns its keep? and does Student-t beat Normal?
for nm, idata in [('Normal+FE', id_normal), ('Normal no-FE', id_nofe), ('Student-t+FE', id_studt)]:
    lo = az.loo(idata, pointwise=True)
    print(f"{nm:14s} elpd={float(lo.elpd):7.1f}  p_loo={float(lo.p):6.1f}  bad_k(>{float(lo.good_k):.2f})={int((lo.pareto_k.values>float(lo.good_k)).sum())}")
print()
print(az.compare({'Normal+FE': id_normal, 'no_FE': id_nofe})[['rank','elpd','elpd_diff','dse']].to_string())
print()
print(az.compare({'Student-t': id_studt, 'Normal': id_normal})[['rank','elpd','elpd_diff','dse']].to_string())
print(f"\nStudent-t nu = {id_studt.posterior['nu'].values.mean():.1f}  (small => very heavy tails)")

# what's driving the fat tails?
ext = mk.reindex(mk['d_T'].abs().sort_values(ascending=False).index)
print("\nmost extreme |d_T| (the outliers):")
print(ext[['country','year','co2_per_capita','d_T']].head(6).to_string(index=False))
print(f"median co2/cap at top-20 outliers: {ext.head(20)['co2_per_capita'].median():.2f}  vs sample: {mk['co2_per_capita'].median():.2f}")

Normal+FE      elpd= 5521.3  p_loo= 149.6  bad_k(>0.70)=0


Normal no-FE   elpd= 5294.4  p_loo=   8.1  bad_k(>0.70)=0


Student-t+FE   elpd= 6085.0  p_loo= 205.3  bad_k(>0.70)=0



           rank    elpd  elpd_diff   dse
Normal+FE     0  5500.0        0.0   0.0
no_FE         1  5300.0     -230.0  25.0



           rank    elpd  elpd_diff   dse
Student-t     0  6100.0        0.0   0.0
Normal        1  5500.0     -600.0  54.0

Student-t nu = 2.2  (small => very heavy tails)

most extreme |d_T| (the outliers):
    country  year  co2_per_capita       d_T
       Laos  2014           0.654  0.484279
       Laos  2013           0.635  0.439709
      Yemen  2014           0.880 -0.384100
   Mongolia  2010           5.098  0.365134
      Yemen  2013           0.931 -0.363352
Afghanistan  2006           0.085  0.336129
median co2/cap at top-20 outliers: 0.43  vs sample: 2.60


**Validation read.**
1. **PPC: Normal fails the tails.** Center fits (mean/sd/q95 p≈0.5) but tails don't (q05≈0.00, min≈1.0, max≈0.00) — the Normal can't reach the extremes and smears the core to compensate.
2. **FE earns its keep predictively:** Normal+FE beats no-FE by `elpd_diff ≈ −230` (dse ≈ 25, ~9σ) — our causal architecture also predicts held-out data far better. Zero bad Pareto-k → LOO reliable.
3. **The outliers are tiny-emission / conflict controls** (Laos, Yemen, Afghanistan, Sierra Leone — median co2/cap ≈ 0.4 vs 2.6). Huge log-swings off a tiny base; none are tax/ETS countries.
4. **Student-t fixes it and is decisively better:** LOO `elpd` jumps ~+560 (dse ~54), estimated `nu ≈ 2` (very heavy tails). It robustly *down-weights* the noisy controls instead of letting them inflate `sigma`.

**Consequence:** under the robust likelihood the effects **attenuate ~25%** (a good robustness result — they're no longer outlier-driven) but the story is unchanged. We adopt Student-t for the engine.

In [29]:
# ROBUST ENGINE: rebuild all 4 Kaya channels with a Student-t likelihood
def fit_engine_t(yvals):
    with pm.Model():
        alpha = pm.Normal('mu_a', 0, 0.5) + pm.HalfNormal('sigma_a', 0.2) * pm.Normal('z_a', 0, 1, shape=nc)
        gamma = pm.HalfNormal('sigma_g', 0.1) * pm.Normal('z_g', 0, 1, shape=ny)
        mu_tax  = pm.Normal('mu_tax', 0, 0.2); tau_tax = pm.HalfNormal('tau_tax', 0.2)
        beta_tax = pm.Deterministic('beta_tax', mu_tax + tau_tax * pm.Normal('z_tax', 0, 1, shape=nc))
        mu_ets  = pm.Normal('mu_ets', 0, 0.2); tau_ets = pm.HalfNormal('tau_ets', 0.2)
        beta_ets = pm.Deterministic('beta_ets', mu_ets + tau_ets * pm.Normal('z_ets', 0, 1, shape=nc))
        delta = pm.Normal('delta', 0, 0.2); sigma = pm.HalfNormal('sigma', 0.2)
        nu = pm.Gamma('nu', 2, 0.1)
        mu_i = alpha[c_idx] + gamma[t_idx] + beta_tax[c_idx]*tax_k + beta_ets[c_idx]*ets_k + delta*both_k
        pm.StudentT('y_obs', nu=nu, mu=mu_i, sigma=sigma, observed=yvals)
        idata = pm.sample(draws=1000, tune=1000, chains=4, target_accept=0.95,
                          random_seed=42, nuts_sampler='nutpie', progressbar=False)
    return (idata.posterior['beta_tax'].values.reshape(-1, nc),
            idata.posterior['beta_ets'].values.reshape(-1, nc),
            idata.posterior['delta'].values.flatten(),
            float(idata.posterior['nu'].values.mean()))

bt, be, dl = {}, {}, {}
for c in CH:
    bt[c], be[c], dl[c], nu = fit_engine_t(mk[f'd_{c}'].values)
    print(f"fit channel {c}  nu={nu:.1f}")

NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma, nu]


C:\Users\willn\OneDrive\Documents\github\CarbonTaxModellindAndRecommendation\.venv\Lib\site-packages\pytensor\tensor\rewriting\elemwise.py:1034: UserWarning: Loop fusion failed because the resulting node would exceed the kernel argument limit.
  warn(


fit channel T  nu=2.2


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma, nu]


C:\Users\willn\OneDrive\Documents\github\CarbonTaxModellindAndRecommendation\.venv\Lib\site-packages\pytensor\tensor\rewriting\elemwise.py:1034: UserWarning: Loop fusion failed because the resulting node would exceed the kernel argument limit.
  warn(


fit channel K  nu=1.6


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma, nu]


C:\Users\willn\OneDrive\Documents\github\CarbonTaxModellindAndRecommendation\.venv\Lib\site-packages\pytensor\tensor\rewriting\elemwise.py:1034: UserWarning: Loop fusion failed because the resulting node would exceed the kernel argument limit.
  warn(


fit channel I  nu=1.8


NUTS[nutpie]: [mu_a, sigma_a, z_a, sigma_g, z_g, mu_tax, tau_tax, z_tax, mu_ets, tau_ets, z_ets, delta, sigma, nu]


C:\Users\willn\OneDrive\Documents\github\CarbonTaxModellindAndRecommendation\.venv\Lib\site-packages\pytensor\tensor\rewriting\elemwise.py:1034: UserWarning: Loop fusion failed because the resulting node would exceed the kernel argument limit.
  warn(


fit channel A  nu=2.2


In [30]:
def contrast(country, ut, ue, chan):
    p = pos_of[country]
    return bt[chan][:, p]*ut + be[chan][:, p]*ue + dl[chan]*(ut*ue)

def recommend(country, use_tax=True, use_ets=True):
    policy = ' + '.join([p for p, on in [('tax', use_tax), ('ETS', use_ets)] if on]) or 'nothing'
    tot = contrast(country, use_tax, use_ets, 'T')
    info = 'data-informed' if tax_years.get(country, 0) > 0 else 'EXTRAPOLATED (never taxed)'
    print(f"\n=== {country} | adopt {policy} | {info} ===")
    print(f"  TOTAL {tot.mean():+.4f}/yr  90% CI [{np.percentile(tot,5):+.4f}, {np.percentile(tot,95):+.4f}]  "
          f"P(reduce)={(tot<0).mean():.2f}  (~{tot.mean()*3*100:+.1f}% over 3yr)")
    for c in ['K', 'I', 'A']:
        d = contrast(country, use_tax, use_ets, c)
        sh = 100*d.mean()/tot.mean() if abs(tot.mean()) > 1e-9 else float('nan')
        print(f"    {CH[c]:30s} {d.mean():+.4f}  share~{sh:4.0f}%")

for ut, ue in [(True, True), (False, True), (True, False)]:
    recommend(known, ut, ue)
recommend(never, True, True)


=== Sweden | adopt tax + ETS | data-informed ===
  TOTAL -0.0260/yr  90% CI [-0.0435, -0.0099]  P(reduce)=1.00  (~-7.8% over 3yr)
    fuel-switching (carbon int.)   -0.0144  share~  55%
    efficiency (energy int.)       -0.0050  share~  19%
    activity (affluence)           -0.0134  share~  51%

=== Sweden | adopt ETS | data-informed ===
  TOTAL -0.0271/yr  90% CI [-0.0430, -0.0134]  P(reduce)=1.00  (~-8.1% over 3yr)
    fuel-switching (carbon int.)   -0.0093  share~  34%
    efficiency (energy int.)       +0.0009  share~  -3%
    activity (affluence)           -0.0248  share~  92%

=== Sweden | adopt tax | data-informed ===
  TOTAL -0.0188/yr  90% CI [-0.0321, -0.0067]  P(reduce)=0.99  (~-5.7% over 3yr)
    fuel-switching (carbon int.)   -0.0054  share~  28%
    efficiency (energy int.)       -0.0140  share~  74%
    activity (affluence)           -0.0009  share~   5%

=== Afghanistan | adopt tax + ETS | EXTRAPOLATED (never taxed) ===
  TOTAL -0.0205/yr  90% CI [-0.0437, +0.0027]  

**Robust engine — final read.**

- **Totals are firm and the story is unchanged:** ETS ≈ −0.027 (P=1.00), tax-alone ≈ −0.019 (P=0.99), and **both ≈ ETS alone** — adding a tax on top of ETS does almost nothing. The "ETS carries it, tax redundant" finding *strengthened* under robustness.
- **Magnitudes are ~25% smaller** than the Normal engine — these are the trustworthy, outlier-resistant numbers.
- **Mechanism shares are now directional-only:** robust effects are smaller and the channels are heavy-tailed (`nu` 1.6–2.2), so the percentage split wobbles (e.g. ETS activity share inflates). Read the **totals** as the deliverable; treat the breakdown as a qualitative "where does it mostly flow" rather than precise accounting.

This completes the backend: a **validated, robust** causal recommendation engine. `recommend(country, use_tax, use_ets)` → robust total Δ + 90% CI + P(reduce) + (qualitative) mechanism breakdown, with honest extrapolation flags. Remaining: the **Streamlit dashboard** (build-fast).

## Step 10 — Reference-class (support) diagnostic

The engine learned carbon pricing from **42 countries that are rich and high-governance** (median GDP/cap $32k vs $8k; governance +1.1 vs −0.5 SD). For the other 121 — including *every* high-leverage emitter (China, India, USA, Indonesia) — a recommendation is extrapolation. The `EXTRAPOLATED` flag was binary; this makes it **graded and explained**.

The diagnostic is pure *data* (no model): measure how far a country sits from the priced countries' covariate cloud on the two dimensions that matter — **income** (GDP/cap) and **governance** (implementation capacity, the dominant effect-moderator) — via **Mahalanobis distance**, calibrated against the priced countries' own spread. Labels: DATA-INFORMED (priced) / NEAR-SUPPORT / FAR, with the reason.

In [31]:
# per-country covariates (latest year with both present)
df['log_gdp_pc'] = np.log(df['gdp'] / df['population'])
_cc = ['log_gdp_pc', 'implementation_capacity_z']
_cov = df.dropna(subset=_cc).sort_values('year').groupby('country')[_cc].last()

_ever = df.groupby('country').agg(tax=('has_tax', 'max'), ets=('has_ets', 'max'))
priced_set = _ever[(_ever['tax'] == 1) | (_ever['ets'] == 1)].index
_pin = _cov.index.intersection(priced_set)

_P = _cov.loc[_pin].values
_mu = _P.mean(axis=0)
_Sinv = np.linalg.inv(np.cov(_P, rowvar=False))
_cov['d'] = [float(np.sqrt((x - _mu) @ _Sinv @ (x - _mu))) for x in _cov.values]
_thresh = _cov.loc[_pin, 'd'].quantile(0.90)

def _pctile(country, col):
    return (_cov.loc[_pin, col] < _cov.loc[country, col]).mean() * 100

def reference_class(country):
    if country in priced_set:
        return 'DATA-INFORMED', 'observed adopting carbon pricing'
    if country not in _cov.index:
        return 'UNKNOWN', 'no covariate data'
    reasons = []
    for col, name in [('log_gdp_pc', 'income'), ('implementation_capacity_z', 'governance')]:
        p = _pctile(country, col)
        if p < 1:    reasons.append(f'{name} below observed range')
        elif p > 99: reasons.append(f'{name} above observed range')
        elif p <= 10: reasons.append(f'low-{name} edge')
    tag = 'NEAR-SUPPORT extrap.' if _cov.loc[country, 'd'] <= _thresh else 'FAR extrap.'
    return tag, ('; '.join(reasons) if reasons else 'inside cloud but unobserved')

print(f"reference class: {len(_pin)} priced countries | Mahalanobis 90th-pct threshold = {_thresh:.2f}")
print(pd.Series({c: reference_class(c)[0] for c in _cov.index}).value_counts().to_string())

reference class: 42 priced countries | Mahalanobis 90th-pct threshold = 2.37
FAR extrap.             97
DATA-INFORMED           42
NEAR-SUPPORT extrap.    24


In [32]:
def recommend(country, use_tax=True, use_ets=True):
    policy = ' + '.join([p for p, on in [('tax', use_tax), ('ETS', use_ets)] if on]) or 'nothing'
    tag, why = reference_class(country)
    tot = contrast(country, use_tax, use_ets, 'T')
    print(f"\n=== {country} | adopt {policy} ===")
    print(f"  CONFIDENCE: {tag}  ({why})")
    print(f"  TOTAL {tot.mean():+.4f}/yr  90% CI [{np.percentile(tot,5):+.4f}, {np.percentile(tot,95):+.4f}]  "
          f"P(reduce)={(tot<0).mean():.2f}  (~{tot.mean()*3*100:+.1f}% over 3yr)")
    for c in ['K', 'I', 'A']:
        d = contrast(country, use_tax, use_ets, c)
        sh = 100 * d.mean() / tot.mean() if abs(tot.mean()) > 1e-9 else float('nan')
        print(f"    {CH[c]:30s} {d.mean():+.4f}  share~{sh:4.0f}%")

# the gradation in actual engine output: Tier-1 -> near -> far
for ctry in ['Sweden', 'United States', 'China', 'India', 'Afghanistan']:
    if ctry in pos_of:
        recommend(ctry, True, True)


=== Sweden | adopt tax + ETS ===
  CONFIDENCE: DATA-INFORMED  (observed adopting carbon pricing)
  TOTAL -0.0260/yr  90% CI [-0.0435, -0.0099]  P(reduce)=1.00  (~-7.8% over 3yr)
    fuel-switching (carbon int.)   -0.0144  share~  55%
    efficiency (energy int.)       -0.0050  share~  19%
    activity (affluence)           -0.0134  share~  51%

=== United States | adopt tax + ETS ===
  CONFIDENCE: NEAR-SUPPORT extrap.  (inside cloud but unobserved)
  TOTAL -0.0207/yr  90% CI [-0.0442, +0.0028]  P(reduce)=0.93  (~-6.2% over 3yr)
    fuel-switching (carbon int.)   -0.0124  share~  60%
    efficiency (energy int.)       -0.0048  share~  23%
    activity (affluence)           -0.0048  share~  23%

=== China | adopt tax + ETS ===
  CONFIDENCE: NEAR-SUPPORT extrap.  (low-governance edge)
  TOTAL -0.0208/yr  90% CI [-0.0439, +0.0027]  P(reduce)=0.93  (~-6.2% over 3yr)
    fuel-switching (carbon int.)   -0.0125  share~  60%
    efficiency (energy int.)       -0.0052  share~  25%
    activity 

**Read.** The engine now states *why* to trust (or not trust) each recommendation:
- **Sweden** → DATA-INFORMED: tight CI, take it at face value.
- **United States** → NEAR-SUPPORT: rich and reasonably governed, just never priced — moderate confidence.
- **China / Indonesia** → NEAR-SUPPORT but flagged *low-income / low-governance edge* — at the boundary of what we've observed.
- **India / Afghanistan** → FAR: income/governance *below the entire observed range* — the point estimate is a population-prior guess, not knowledge.

This is the honest core of the engine: **42 data-informed, 24 near-support, 97 far** out of 163. It doesn't refuse to answer for unobserved countries — it answers, then tells you the answer is extrapolation and on which dimension. The dashboard should lead with this confidence tag + CI, not the point estimate.

**Backend now complete and self-aware.** Remaining Phase 3: the Streamlit dashboard (build-fast).